cmd/gator/test/test.go

```diff
@@ -96,31 +102,64 @@ func run(cmd *cobra.Command, args []string) {
 	os.Exit(exitCode)
 }
 
-func formatOutput(flagOutput string, results []*test.GatorResult) string {
+func formatOutput(flagOutput string, results []*test.GatorResult, stats []*instrumentation.StatsEntry) string {
 	switch strings.ToLower(flagOutput) {
 	case stringJSON:
-		b, err := json.MarshalIndent(results, "", "    ")
-		if err != nil {
-			errFatalf("marshaling validation json results: %v", err)
+		var jsonB []byte
+		var err error
+		if stats != nil {
+			statsAndResults := map[string]interface{}{"results": results, "stats": stats}
+			jsonB, err = json.MarshalIndent(statsAndResults, "", fourSpaceTab)
+			if err != nil {
+				cmdutils.ErrFatalf("marshaling validation json results and stats: %v", err)
+			}
+		} else {
+			jsonB, err = json.MarshalIndent(results, "", fourSpaceTab)
+			if err != nil {
+				cmdutils.ErrFatalf("marshaling validation json results: %v", err)
+			}
 		}
-		return string(b)
+
+		return string(jsonB)
 	case stringYAML:
 		yamlResults := test.GetYamlFriendlyResults(results)
 		jsonb, err := json.Marshal(yamlResults)
 		if err != nil {
-			errFatalf("pre-marshaling results to json: %v", err)
+			cmdutils.ErrFatalf("pre-marshaling results to json: %v", err)
 		}
 
 		unmarshalled := []*test.YamlGatorResult{}
 		err = json.Unmarshal(jsonb, &unmarshalled)
 		if err != nil {
-			errFatalf("pre-unmarshaling results from json: %v", err)
+			cmdutils.ErrFatalf("pre-unmarshaling results from json: %v", err)
 		}
 
-		yamlb, err := yaml.Marshal(unmarshalled)
-		if err != nil {
-			errFatalf("marshaling validation yaml results: %v", err)
+		var yamlb []byte
+		if stats != nil {
+			statsAndResults := map[string]interface{}{"results": yamlResults, "stats": stats}
+
+			statsJSONB, err := json.Marshal(stats)
```

Comment:
What is the purpose of round-tripping stats here? It doesn't look like `unmarshalledStats` gets used?

Reply:
yeah `unmarshalledStats` is not used directly but I thought we wanted to make sure that we can marshal/ unmarshal the stats.

So `statsAndResults` has two fields: `{"results": yamlResults, "stats": stats}` we already know that yamlResults marshalls/ unmarshalls from the lines L124-135 so we do a similar check here for the stats. If both `yamlResults` and `stats` can marshall/ unmarshall, then `statsAndResults` should not have an issue.

What do you think of this? Or am I missing something simpler? I could create a special type for `statsAndResults` but I figured that may be the more inelegant solution

Reply:
Why do we have to marshal `stats` and `yaml` results separately? 

`Marshal` should traverse recursively. Wouldn't marshaling `statsAndResults` achieve the same thing?

Reply:
> Why do we have to marshal stats and yaml results separately?

Originally, we round trip json to weed out any marshalling issues, before roundtripping yaml. I took a stab at answering why I did this separately for results and stats in the previous comment:

> So statsAndResults has two fields: {"results": yamlResults, "stats": stats} we already know that yamlResults marshalls/ unmarshalls from the lines L124-135 so we do a similar check here for the stats. If both yamlResults and stats can marshall/ unmarshall, then statsAndResults should not have an issue.

Since it seems like this is causing some confusion, I will anonymously create a type for single use there as I was suggesting instead:

> I could create a special type for statsAndResults but I figured that may be the less straightforward solution.

Reply:
Let me know if this is not clearer or cleaner folks; Feel free to comment or reopen this comment.